In [1]:
import numpy as np

arr = np.arange(180).reshape(3, 3, 4, 5)

print("arr:\n", arr)

print("\niterating by rows:")
for row in arr:
    print(row)

arr:
 [[[[  0   1   2   3   4]
   [  5   6   7   8   9]
   [ 10  11  12  13  14]
   [ 15  16  17  18  19]]

  [[ 20  21  22  23  24]
   [ 25  26  27  28  29]
   [ 30  31  32  33  34]
   [ 35  36  37  38  39]]

  [[ 40  41  42  43  44]
   [ 45  46  47  48  49]
   [ 50  51  52  53  54]
   [ 55  56  57  58  59]]]


 [[[ 60  61  62  63  64]
   [ 65  66  67  68  69]
   [ 70  71  72  73  74]
   [ 75  76  77  78  79]]

  [[ 80  81  82  83  84]
   [ 85  86  87  88  89]
   [ 90  91  92  93  94]
   [ 95  96  97  98  99]]

  [[100 101 102 103 104]
   [105 106 107 108 109]
   [110 111 112 113 114]
   [115 116 117 118 119]]]


 [[[120 121 122 123 124]
   [125 126 127 128 129]
   [130 131 132 133 134]
   [135 136 137 138 139]]

  [[140 141 142 143 144]
   [145 146 147 148 149]
   [150 151 152 153 154]
   [155 156 157 158 159]]

  [[160 161 162 163 164]
   [165 166 167 168 169]
   [170 171 172 173 174]
   [175 176 177 178 179]]]]

iterating by rows:
[[[ 0  1  2  3  4]
  [ 5  6  7  8  9]
  [10 11 12 1

In [2]:
arr_1 = arr.reshape(180)
print(arr_1)

[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179]


In [3]:
conv_0 = arr[:,:,:,0]
print("conv_0:\n", conv_0)
# conv kernel
conv_0_d_0 = arr[:,:,0,0]
conv_0_d_1 = arr[:,:,1,0]
conv_0_d_2 = arr[:,:,2,0]
conv_0_d_3 = arr[:,:,3,0]

conv_1_d_0 = arr[:,:,0,1]
conv_1_d_1 = arr[:,:,1,1]
conv_1_d_2 = arr[:,:,2,1]

# 第0,1,2列
conv_0_d_0_c_0 = arr[:,0,0,0]
conv_0_d_0_c_1 = arr[:,1,0,0]
conv_0_d_0_c_2 = arr[:,2,0,0]
# 第0,1,2行
conv_0_d_0_r_0 = arr[0, :,0,0]
conv_0_d_0_r_1 = arr[1, :,0,0]
conv_0_d_0_r_2 = arr[2, :,0,0]

# 一些元素(0,0),(0,1),(0.2)
conv_0_d_0_c_0_r_0 = arr[0,0,0,0]
conv_0_d_0_c_1_r_0 = arr[0,1,0,0]
conv_0_d_0_c_2_r_0 = arr[0,2,0,0]

print("conv_0_d_0:\n", conv_0_d_0)
print("conv_0_d_1:\n", conv_0_d_1)
print("conv_0_d_2:\n", conv_0_d_2)
print("conv_0_d_3:\n", conv_0_d_3)

print("conv_0_d_0_c_0:\n", conv_0_d_0_c_0)
print("conv_0_d_0_c_1:\n", conv_0_d_0_c_1)
print("conv_0_d_0_c_2:\n", conv_0_d_0_c_2)

print("conv_0_d_0_r_0:\n", conv_0_d_0_r_0)
print("conv_0_d_0_r_1:\n", conv_0_d_0_r_1)
print("conv_0_d_0_r_2:\n", conv_0_d_0_r_2)

print("conv_0_d_0:(0,0),(0,1),(0,2):\n", 
     conv_0_d_0_c_0_r_0, conv_0_d_0_c_1_r_0,
     conv_0_d_0_c_2_r_0)
# open file, "a"-> append, "w" -> overwrite
file=open("coefficient_file/4d_ndarray_test.txt", "w")
file.write("1st conv kernel:\n")
np.savetxt(file,conv_0_d_0)
np.savetxt(file,conv_0_d_1)
np.savetxt(file,conv_0_d_2)
file.write("2nd conv kernel:\n")
np.savetxt(file,conv_1_d_0)
np.savetxt(file,conv_1_d_1)
np.savetxt(file,conv_1_d_2)
file.close()

conv_0:
 [[[  0   5  10  15]
  [ 20  25  30  35]
  [ 40  45  50  55]]

 [[ 60  65  70  75]
  [ 80  85  90  95]
  [100 105 110 115]]

 [[120 125 130 135]
  [140 145 150 155]
  [160 165 170 175]]]
conv_0_d_0:
 [[  0  20  40]
 [ 60  80 100]
 [120 140 160]]
conv_0_d_1:
 [[  5  25  45]
 [ 65  85 105]
 [125 145 165]]
conv_0_d_2:
 [[ 10  30  50]
 [ 70  90 110]
 [130 150 170]]
conv_0_d_3:
 [[ 15  35  55]
 [ 75  95 115]
 [135 155 175]]
conv_0_d_0_c_0:
 [  0  60 120]
conv_0_d_0_c_1:
 [ 20  80 140]
conv_0_d_0_c_2:
 [ 40 100 160]
conv_0_d_0_r_0:
 [ 0 20 40]
conv_0_d_0_r_1:
 [ 60  80 100]
conv_0_d_0_r_2:
 [120 140 160]
conv_0_d_0:(0,0),(0,1),(0,2):
 0 20 40


卷积核0，channel 0

conv_0_d_0:

 [[  0  20  40]
  [ 60  80 100]
  [120 140 160]]
  
卷积核0，channel 1

conv_0_d_1:

 [[  5  25  45]
 [ 65  85 105]
 [125 145 165]]
 
卷积核0，channel 2

conv_0_d_2:

 [[ 10  30  50]
 [ 70  90 110]
 [130 150 170]]
 
卷积核0，channel 3

conv_0_d_3:

 [[ 15  35  55]
 [ 75  95 115]
 [135 155 175]]

In [12]:
print("4d array :")
print("    type:",type(arr))
print("    dtype:", arr.dtype)
print("    shape:", arr.shape)

print(arr.shape[0], arr.shape[1], arr.shape[2], arr.shape[3])
# shape=(filter_height, filter_width, in_channals, out_channels)
# shape=(row of filter, column of filter, number of channals, number of kernel)

4d array :
    type: <class 'numpy.ndarray'>
    dtype: int64
    shape: (3, 3, 4, 5)
3 3 4 5


In [4]:
# write coefiicent by iterating each kernel, 3 parameter each row
file=open("coefficient_file/4d_ndarray_test_1.txt", "w")
for i in range(arr.shape[3]):
    description = str(i+1) + " conv kernel:\n"
    file.write(description)
    for j in range(arr.shape[2]):
        np.savetxt(file,arr[:,:,j,i])
        #for k in range(arr.shape[1]):            
            #for l in range(arr.shape[0]):
                #print("*", end=" ")
file.close()

In [5]:
# write coefiicent by itarating each row of kernel, 1 parameter each row
file=open("coefficient_file/4d_ndarray_test_2.txt", "w")
for i in range(arr.shape[3]):
    description = str(i+1) + " conv kernel:\n"
    file.write(description)
    for j in range(arr.shape[2]):
        for k in range(arr.shape[1]):
            np.savetxt(file,arr[:,k,j,i])
            #for l in range(arr.shape[0]):
                #print("*", end=" ")
file.close()

In [7]:
for i in range(arr.shape[0]):
    for j in range(arr.shape[1]):
        for k in range(arr.shape[2]):
            for l in range(arr.shape[3]):
                print(arr[i, j, k, l], end=" ")
                
# print(arr[i, j, k, l], end=" ")

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 

In [ ]:
#import numpy 
#numpy.savetxt("coefficient_file/w__test_1.csv", arr, delimiter=",")
# ValueError: Expected 1D or 2D array, got 4D array instead

In [ ]:
# write coefiicent by iterating each row of kernels, 1 parameter each row
#file=open("coefficient_file/w__test_1.txt", "w")
# we test default itrating order for ndarray
#np.savetxt(file, arr[:, :, :, :])
#file.close()
# ValueError: Expected 1D or 2D array, got 4D array instead

In [ ]:
#import pandas as pd
#pd.DataFrame(arr).to_csv("coefficient_file/w__test_1.csv")
# ValueError: Must pass 2-d input

In [14]:
arr_2 = np.arange(200).reshape(25, 8)
print("2d array :")
print("    type:",type(arr_2))
print("    dtype:", arr_2.dtype)
print("    shape:", arr_2.shape)

print("arr_2:\n", arr_2)

2d array :
    type: <class 'numpy.ndarray'>
    dtype: int64
    shape: (25, 8)
arr_2:
 [[  0   1   2   3   4   5   6   7]
 [  8   9  10  11  12  13  14  15]
 [ 16  17  18  19  20  21  22  23]
 [ 24  25  26  27  28  29  30  31]
 [ 32  33  34  35  36  37  38  39]
 [ 40  41  42  43  44  45  46  47]
 [ 48  49  50  51  52  53  54  55]
 [ 56  57  58  59  60  61  62  63]
 [ 64  65  66  67  68  69  70  71]
 [ 72  73  74  75  76  77  78  79]
 [ 80  81  82  83  84  85  86  87]
 [ 88  89  90  91  92  93  94  95]
 [ 96  97  98  99 100 101 102 103]
 [104 105 106 107 108 109 110 111]
 [112 113 114 115 116 117 118 119]
 [120 121 122 123 124 125 126 127]
 [128 129 130 131 132 133 134 135]
 [136 137 138 139 140 141 142 143]
 [144 145 146 147 148 149 150 151]
 [152 153 154 155 156 157 158 159]
 [160 161 162 163 164 165 166 167]
 [168 169 170 171 172 173 174 175]
 [176 177 178 179 180 181 182 183]
 [184 185 186 187 188 189 190 191]
 [192 193 194 195 196 197 198 199]]


In [17]:
# store row fistly, then column , storage format is correct
file=open("coefficient_file/2d_ndarray_test.txt", "w")
np.savetxt(file, arr_2[:, :], delimiter="\n")
file.close()

In [18]:
file=open("coefficient_file/2d_ndarray_test_1.txt", "w")
# iterating by row
for i in range(arr_2.shape[0]):
    np.savetxt(file, arr_2[i, :])
file.close()